closed/src/jdk.crypto.ec/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -224,11 +210,14 @@ protected int engineGenerateSecret(byte[] sharedSecret, int offset)
         long nativePublicKey = this.publicKey.getNativePtr();
         long nativePrivateKey = this.privateKey.getNativePtr();
         if ((nativePublicKey == -1) || (nativePrivateKey == -1)) {
-            if (curveSupported.putIfAbsent(this.curve, Boolean.FALSE) != null) {
-                throw new ProviderException("Could not convert keys to native format");
+            absent = NativeECUtil.putCurveIfAbsent(this.curve, Boolean.FALSE);
+            /* only print the first time a curve is used */
+            if (absent && nativeCryptTrace) {
+                System.err.println(this.curve +
+                        " is not supported by OpenSSL, using Java crypto implementation for preparing agreement.");
             }
-            if (nativeCryptTrace) {
-                System.err.println(this.curve + " is not supported by OpenSSL, using Java crypto implementation.");
+            if (!absent) {
+                throw new ProviderException("Could not convert keys to native format");
             }
```

Comment:
Suggest reordering this (and similar sequences below) so `absent` need only be tested once:
```java
            absent = NativeECUtil.putCurveIfAbsent(this.curve, Boolean.FALSE);
            if (!absent) {
                throw new ProviderException("Could not convert keys to native format");
            }
            /* only print the first time a curve is used */
            if (nativeCryptTrace) {
                System.err.println(this.curve +
                        " is not supported by OpenSSL, using Java crypto implementation for preparing agreement.");
            }
```

Reply:
Changed it everywhere.